Implemented middle click to jump-to-location

ChaoticMind opened the pull request:
Prerequisite for #290. 

This only covers the middle click equivalent to clicking on the minimap (point 1 in the original comment of #290). Dragging is not yet implemented. 

In the current implementation, middle clicking (even on this "visibleArea") centers the visible area to that destination. This may or may not prove awkward after dragging is implemented. I thought it would be more consistent this way. However, it looks like most scroll bar implementations behave the same way if they're left clicked or middle clicked (if started dragging from the bar itself), so I may revisit this while implementing dragging.

ChaoticMind commented:
Do you have any idea why [this](https://github.com/ChaoticMind/minimap/commit/4db8e8dc1d93ab052abaaf3e87fa679825db2340#diff-0bb5bdcf61dee1a981472763ac09de4aR31) doesn't seem to work?

fundon commented:
@ChaoticMind I will test it.
And thank you for your PR. :+1:

fundon commented:
Hi @ChaoticMind
Could you give me a screenshot(gif)?
I have not a middle-button. Sorry for that.

abe33 commented:
Hi @ChaoticMind, thanks for taking care of that. 

The reason it doesn't work is because you pass the `btn` as second argument, which can't work as there's 4 more parameter to pass before that.

Also, I'm not sure to understand why you toggle a setting on middle click, maybe it was just a test, but it shouldn't work that way.

I'll take some time this evening to clean the event helpers a bit so that we can pass the which property in the helpers.

ChaoticMind commented:
I'm actually going away for a couple of weeks, so if you want to hold off on merging this until the rest of the commits/fixups, that's fine. 
I can reopen when I'm back. Sorry about that.

ChaoticMind commented:
@fundon Here's an animated gif: ![middle-click preview](https://fat.gfycat.com/GroundedGiddyBunny.gif)

ChaoticMind commented:
@abe33 Yes, the setting toggling was indeed just a test.

I have a couple of questions: 
In [this](https://github.com/atom-minimap/minimap/blob/v4.7.6/spec/minimap-element-spec.coffee) spec file:
- Why is canvas.offsetHeight (810) different than "height" (405) obtained from {height} = canvas.getBoundingClientRect()? 
For example, if one adds to the "scrolls the editor to the line below the mouse" test the following lines:
```coffeescript
  console.log(canvas.offsetHeight)
  {height} = canvas.getBoundingClientRect()
  console.log(height)
```

- Why is canvas's 'height' (obtained from getBoundingClientRect) (405) larger than editor.getHeight() (400)? I had to get their min in one of the tests to get rid of the discrepancy.

I rebased this PR and improved the old tests. I added a commit to support to click-and-drag behavior. Two issues remain however before this can be merged:
1. I used a workaround [here](https://github.com/ChaoticMind/minimap/commit/86c155b7996193b52c263f155921b3cb76ac13e0#diff-522976b2d1fa347c17ee996341ef550bR460) which artificially sets pageY so that the dragOffset will be correct. I'm not sure why in "startDrag()", the call to [visibleArea.getBoundingClientRect](https://github.com/ChaoticMind/minimap/blob/86c155b7996193b52c263f155921b3cb76ac13e0/lib/minimap-element.coffee#L516) gets the old coords of visibleArea (pre-middle click). Maybe it's something with the callback order, but I couldn't figure it out. I tried [forcing](https://github.com/ChaoticMind/minimap/commit/86c155b7996193b52c263f155921b3cb76ac13e0#diff-522976b2d1fa347c17ee996341ef550bR458) an update to no avail.
2. I have a reference error [here](https://github.com/ChaoticMind/minimap/commit/86c155b7996193b52c263f155921b3cb76ac13e0#diff-d279fd0a1035f9f0d4c9f42bb0dbf4dcR366) in the latest test that I couldn't figure out.

I think this would be ready to be merged after these two issues are resolved.

abe33 commented:
Nice job, just make sure Travis tests pass with the previous values (see my comments on #312) and it'll be good for me.

abe33 commented:
@ChaoticMind 

> Why is canvas.offsetHeight (810) different than "height" (405) obtained from {height} = canvas.getBoundingClientRect()?

`getBoundingClientRect` includes css transforms in the computed bounds when `offsetHeight` doesn't, so if you have a retina screen the canvas is twice as tall as the on-screen size.

See what I get on my iMac (which is not a retina screen): 
![common_sass_-__users_cedric_development_ruby_sumo_-_atom](https://cloud.githubusercontent.com/assets/247140/7067052/a448fd64-dec8-11e4-8154-36e2c2c60a2c.png)

> Why is canvas's 'height' (obtained from getBoundingClientRect) (405) larger than editor.getHeight() (400)? I had to get their min in one of the tests to get rid of the discrepancy.

The canvas is taller by one line's height in comparison with the editor, this is required as the canvas is translated by a portion of a line's height when scrolling (this is to ensure smooth scrolling as we're only rendering complete lines on the canvas), that way there's no missing pixels at the bottom of the canvas. If we were using exactly the same height, when the minimap is scrolled such as the first line is partially hidden, the amount of negative offset on the top would be missing at the bottom.

As for your two last point, I haven't an answer right now, I'll check that after work.

ChaoticMind commented:
> `getBoundingClientRect` includes css transforms in the computed bounds when offsetHeight doesn't, > so if you have a retina screen the canvas is twice as tall as the on-screen size.

Aha, I see. It's also the same for me in the console, but they're different in the test suite [which is good].

Re: the canvas being taller - it makes sense - however, in the spec suite, they are different by 5 pixels (405 vs 400) but the line height `editor.displayBuffer.getLineHeightInPixels()` is 14 px. It doesn't really matter though, as this is getting pretty off-topic. Thanks for the clarifications!

abe33 commented:
> Re: the canvas being taller - it makes sense - however, in the spec suite, they are different by 5 pixels (405 vs 400) but the line height editor.displayBuffer.getLineHeightInPixels() is 14 px 

We're talking about minimap lines here, whch are obviously smaller than the editor one. In the test suite I use lines of 5px (4 + 1 if I remember correctly). Hence the 5px dif ;)

Also, I've seen that you have opened many issues, thanks for this, I'll just need some time to review them as I was away from the internets this week-end.

abe33 commented:
Sorry for the late reply, it's good to me, I'll publish it soon.
